lib/plugins/stackoverflow/post.dart

```diff
@@ -32,11 +35,92 @@ class StackOverflowPostOwner with _$StackOverflowPostOwner {
       _$StackOverflowPostOwnerFromJson(json);
 }
 
-String postToDiscordMarkdown(StackOverflowPost post) {
-  return '# ${post.title}\n'
+Future<String> postToDiscordMarkdown(
+  StackOverflowPost post,
+  Environment env,
+) async {
+  String header = '# ${post.title}\n'
       '[${post.owner.displayName}](<${post.owner.link}>) - '
       '${timeago.format(DateTime.fromMillisecondsSinceEpoch(post.creationDate * 1000))} | '
       '${post.score} votes | '
-      '[Source](<${post.link}>)\n\n'
-      '${post.bodyMarkdown}';
+      '[Source](<${post.link}>)\n\n';
+
+  String bodyMarkdown = env.pastebinApiKey == null
+      ? post.bodyMarkdown
+      : await shortenWithPastebin(
+          post.bodyMarkdown,
+          PastebinClient(env.pastebinApiKey!),
+          kMaxMessageLength - header.length,
+        );
+
+  String result = '$header$bodyMarkdown';
+
+  if (result.length > kMaxMessageLength) {
+    return '${result.substring(0, kMaxMessageLength - 3)}...';
+  }
+
+  return result;
+}
+
+// Matches code blocks fenced with ```code``` in group 1 and code blocks
+// indented by four spaces in group 2 (SO seems to use the latter quite often).
+final codeblockPattern = RegExp(
+  r'(?:```(?:dart|)\n((?:.*\n)+?)```)|(?:( {4}.+\n(?:(?: {4}.+\n|\n)*) {4}.+)\n)',
+  caseSensitive: false,
+);
+
+Future<String> shortenWithPastebin(
+  String content,
+  PastebinClient client,
+  int maxLength,
+) async {
+  content = content.replaceAll('\r\n', '\n');
+
+  final allMatches = codeblockPattern.allMatches(content);
+  // (start, length, content)
+  final blocks = List.of(allMatches.map((m) => (
+        m.start,
+        m.end - m.start,
+        m.group(1) ?? m.group(2)!,
+      )))
+    // Ignore code blocks less than 200 characters long.
+    ..removeWhere((element) => element.$2 < 200);
+
+  for (final block in blocks) {
+    // Cutting codeblocks that will be cut anyway has no effect
+    if (block.$1 > maxLength) break;
+    // We've managed to cut out enough blocks to make the content fit
+    if (content.length <= maxLength) break;
+
+    final paste = await client.upload(block.$3, language: 'dart');
+
+    final oldLength = content.length;
+
+    String cutPreview = block.$3.split('\n').take(3).join('\n');
```

Comment:
![image](https://github.com/clragon/swan/assets/11785085/0b7f5486-8cad-48b4-8665-cb253340f81b)
![image](https://github.com/clragon/swan/assets/11785085/fb4db227-3dd5-4435-a0ba-fb08b20c6fd7)
I guess we have to apply `trim` as well